src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs

```diff
@@ -135,76 +135,112 @@ internal Memory<byte> GetMemory(int sizeHint)
 
             lock (_sync)
             {
-                BufferSegment segment = _writingHead ?? AllocateWriteHeadUnsynchronized(sizeHint);
+                AllocateWriteHeadUnsynchronized(sizeHint);
+            }
 
-                int bytesLeftInBuffer = segment.WritableBytes;
+            // Slice the AvailableMemory to the WritableBytes size
+            int end = _writingHead.End;
+            Memory<byte> availableMemory = _writingHead.AvailableMemory;
+            availableMemory = availableMemory.Slice(end);
+            return availableMemory;
+        }
 
-                // If inadequate bytes left or if the segment is readonly
-                if (bytesLeftInBuffer == 0 || bytesLeftInBuffer < sizeHint || segment.ReadOnly)
-                {
-                    BufferSegment nextSegment = CreateSegmentUnsynchronized();
-                    nextSegment.SetMemory(_pool.Rent(GetSegmentSize(sizeHint)));
+        internal Span<byte> GetSpan(int sizeHint)
+        {
+            if (_writerCompletion.IsCompleted)
+            {
+                ThrowHelper.ThrowInvalidOperationException_NoWritingAllowed();
+            }
 
-                    segment.SetNext(nextSegment);
+            if (sizeHint < 0)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.minimumSize);
+            }
 
-                    _writingHead = nextSegment;
-                }
+            lock (_sync)
+            {
+                AllocateWriteHeadUnsynchronized(sizeHint);
             }
 
             // Slice the AvailableMemory to the WritableBytes size
             int end = _writingHead.End;
-            Memory<byte> availableMemory = _writingHead.AvailableMemory;
-            availableMemory = availableMemory.Slice(end, availableMemory.Length - end);
+            Span<byte> availableMemory = _writingHead.AvailableMemory.Span;
```

Comment:
nit: `availableMemory` => `availableSpan`

Reply:
Don't use `_writingHead` outside the lock

Reply:
Was used previously outside lock?

Reply:
Good time to stop :)

Reply:
Or wait, it's not supposed to be syncronised.